File: Barchart.BinarySerializer/Types/BinarySerializerUShort.cs
#region Using Statements

using Barchart.BinarySerializer.Buffers;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) ushort values to (and from) a binary data source.
/// </summary>
public class BinarySerializerUShort : IBinaryTypeSerializer<ushort>
{
    #region Constants
        
    private const int ENCODED_LENGTH_IN_BYTES = sizeof(ushort);
        
    #endregion

    #region Methods

    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, ushort value)
    {
        writer.WriteBytes(BitConverter.GetBytes(value));
    }

    /// <inheritdoc />
    public ushort Decode(IDataBufferReader reader)
    {
        return BitConverter.ToUInt16(reader.ReadBytes(ENCODED_LENGTH_IN_BYTES));
    }
        
    /// <inheritdoc />
    public bool GetEquals(ushort a, ushort b)
    {
        return a.Equals(b);
    }

    #endregion
}

File: typescript/src/types/binary-serializer-ushort.ts
import { DataReader } from "../buffers/data-reader.interface";
import { DataWriter } from "../buffers/data-writer.interface";
import { BinaryTypeSerializer } from "./binary-type-serializer.interface";

/**
 * Reads (and writes) ushort values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<number>}
 */
export class BinarySerializerUShort implements BinaryTypeSerializer<number> {
    get sizeInBytes(): number {
        return 2;
    }

    encode(writer: DataWriter, value: number): void {
        const buffer = new ArrayBuffer(this.sizeInBytes);
        const view = new DataView(buffer);
        view.setUint16(0, value, true);
        
        writer.writeBytes(new Uint8Array(buffer));
    }

    decode(reader: DataReader): number {
        const valueBytes = reader.readBytes(this.sizeInBytes);

        return new DataView(valueBytes.buffer, valueBytes.byteOffset, valueBytes.byteLength).getUint16(0, true);
    }

    getEquals(a: number, b: number): boolean {
        return a === b;
    }
}